transport/awssqs/consumer.go

```diff
@@ -0,0 +1,242 @@
+// endpoint is invoked, but before anything is published to the reply.
+func ConsumerAfter(after ...ConsumerResponseFunc) ConsumerOption {
+	return func(c *Consumer) { c.after = append(c.after, after...) }
+}
+
+// ConsumerErrorEncoder is used to encode errors to the consumer reply
+// whenever they're encountered in the processing of a request. Clients can
+// use this to provide custom error formatting. By default,
+// errors will be published with the DefaultErrorEncoder.
+func ConsumerErrorEncoder(ee ErrorEncoder) ConsumerOption {
+	return func(c *Consumer) { c.errorEncoder = ee }
+}
+
+// ConsumerVisbilityTimeOutFunc is used to extend the visibility timeout
+// for messages during when processing them. Clients can
+// use this to provide custom visibility timeout extension. By default,
+// visibility timeout are not extend.
+func ConsumerVisbilityTimeOutFunc(vtFunc VisibilityTimeoutFunc) ConsumerOption {
+	return func(c *Consumer) { c.visibilityTimeoutFunc = vtFunc }
+}
+
+// ConsumerErrorLogger is used to log non-terminal errors. By default, no errors
+// are logged. This is intended as a diagnostic measure. Finer-grained control
+// of error handling, including logging in more detail, should be performed in a
+// custom ConsumerErrorEncoder which has access to the context.
+// Deprecated: Use ConsumerErrorHandler instead.
+func ConsumerErrorLogger(logger log.Logger) ConsumerOption {
+	return func(c *Consumer) { c.errorHandler = transport.NewLogErrorHandler(logger) }
+}
+
+// ConsumerErrorHandler is used to handle non-terminal errors. By default, non-terminal errors
+// are ignored. This is intended as a diagnostic measure. Finer-grained control
+// of error handling, including logging in more detail, should be performed in a
+// custom ConsumerErrorEncoder which has access to the context.
+func ConsumerErrorHandler(errorHandler transport.ErrorHandler) ConsumerOption {
+	return func(c *Consumer) { c.errorHandler = errorHandler }
+}
+
+// ConsumerFinalizer is executed at the end of every request from a publisher through SQS.
+// By default, no finalizer is registered.
+func ConsumerFinalizer(f ...ConsumerFinalizerFunc) ConsumerOption {
+	return func(c *Consumer) { c.finalizer = f }
+}
+
+// Consume calls ReceiveMessageWithContext and handles messages
+// having receiveMsgInput as param allows each user to have his own receive config
+func (c Consumer) Consume(ctx context.Context, receiveMsgInput *sqs.ReceiveMessageInput) error {
+	receiveMsgInput.QueueUrl = c.queueURL
+	out, err := c.sqsClient.ReceiveMessageWithContext(ctx, receiveMsgInput)
+	if err != nil {
+		return err
+	}
+	return c.HandleMessages(ctx, out.Messages)
+}
+
+// HandleMessages handles the consumed messages
+func (c Consumer) HandleMessages(ctx context.Context, msgs []*sqs.Message) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Copy msgs slice in leftMsgs
+	leftMsgs := []*sqs.Message{}
```

Comment:
I actually don't understand what `leftMsgs` is for. Can you explain?

Reply:
When a message is retrieved from SQS, it is made invisible for a certain period, to avoid that two consumers handle the same message concurrently.
When consuming multiple messages at the same time, or for long processing messages, the SDK allows you to extend this visibility timeout.
Once a consumer is done processing a message it should delete the message from the queue using the ReceiptHandle that can be found in the message. If the visibilityTimeout expires so does the ReceiptHandle, this will lead up to message being visible again in the queue and reprocessed.

leftMsgs is used by the consumer's visibilityTimeoutFunc to extend the visibility timeout for the messages that have not been processed yet. 
After a message has been processed and deleted from the queue, leftMsgs should be updated. Those two actions can be done in the consumer's after functions.